api/rest/coordinator.go

```diff
@@ -0,0 +1,256 @@
+	for finishedBundles < numBundles {
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("context canceled before all node bundles finished")
+		case s := <-statuses:
+			if !s.done {
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.id).Info("Got status update. Bundle not ready.")
+				continue
+			}
+			if s.err != nil {
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.id).Warn("Bundle errored")
+				// TODO (https://jira.mesosphere.com/browse/DCOS_OSS-5303): this should be noted in the generated bundle and not just printed in the journal
+				finishedBundles++
+				continue
+			}
+
+			bundlePath := filepath.Join(c.workDir, nodeBundleFilename(s.node))
+			err := c.client.GetFile(ctx, s.node.baseURL, s.id, bundlePath)
+			if err != nil {
+				logrus.WithError(err).WithField("IP", s.node.IP).WithField("ID", s.id).Warn("Could not download file")
+			}
+
+			bundlePaths = append(bundlePaths, bundlePath)
+			finishedBundles++
+		}
+	}
+
+	return mergeZips(bundleID, bundlePaths, c.workDir)
+}
+
+func mergeZips(bundleID string, bundlePaths []string, workDir string) (string, error) {
+
+	bundlePath := filepath.Join(workDir, fmt.Sprintf("bundle-%s.zip", bundleID))
+	mergedZip, err := os.Create(bundlePath)
+	if err != nil {
+		return "", err
+	}
+	defer mergedZip.Close()
+
+	zipWriter := zip.NewWriter(mergedZip)
+	defer zipWriter.Close()
+
+	for _, p := range bundlePaths {
+		err = appendToZip(zipWriter, p)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return mergedZip.Name(), nil
+}
+
+func appendToZip(writer *zip.Writer, path string) error {
+	r, err := zip.OpenReader(path)
+	if err != nil {
+		return fmt.Errorf("could not open %s: %s", path, err)
+	}
+	defer r.Close()
+
+	for _, f := range r.File {
+		rc, err := f.Open()
+		if err != nil {
+			return fmt.Errorf("could not open %s from zip: %s", f.Name, err)
+		}
+
+		file, err := writer.Create(f.Name)
+		if err != nil {
+			return fmt.Errorf("could not create file %s: %s", f.Name, err)
+		}
+		_, err = io.Copy(file, rc)
+		if err != nil {
+			return fmt.Errorf("could not copy file %s to zip: %s", f.Name, err)
+		}
+		rc.Close()
+	}
+
+	return nil
+}
+
+func (c ParallelCoordinator) createBundle(ctx context.Context, node node, id string, jobs chan<- job) BundleStatus {
+	_, err := c.client.CreateBundle(ctx, node.baseURL, id)
+	if err != nil {
+		// Return done status with error. To mark node as errored so file will not be downloaded
+		return BundleStatus{
+			id:   id,
+			node: node,
+			done: true,
+			err:  fmt.Errorf("could not create bundle: %s", err),
+		}
+	}
+
+	// Schedule bundle status check
+	jobs <- func() BundleStatus {
+		return c.waitForDone(ctx, node, id, c.statusCheckInterval, jobs)
+	}
+
+	// Return undone status with no error.
+	return BundleStatus{id: id, node: node}
+}
+
+func (c ParallelCoordinator) waitForDone(ctx context.Context, node node, id string, interval time.Duration, jobs chan<- job) BundleStatus {
```

Comment:
`ParallelCoordinator` already has status check interval we don't need it as an argument
```suggestion
func (c ParallelCoordinator) waitForDone(ctx context.Context, node node, id string, jobs chan<- job) BundleStatus {
```